### ISSUE TITLE
do_request fails when auth.proxy enabled

### ISSUE THREAD

**Issue by atward** [sous-chefs/grafana#247]
Resources that backend to `do_request` fail when using `auth.proxy.enabled = true`

### Scenario:
Grafana is fronted by a reverse proxy, which is controlling authentication. nginx eg:
```
  location /grafana/ {
    proxy_pass http://grafana/;
    proxy_set_header X-WEBAUTH-USER $auth_resp_x_user;
    proxy_set_header Host $host;
    auth_request /validate;
    auth_request_set $auth_resp_x_user $upstream_http_x_user;
  }
```
grafana-cookbook library requests to grafana service are bypassing this authentication by hitting grafana directly. The grafana service however requires the trusted `proxy_header_name` to be supplied. The grafana service treats these as unauthenticated and causes cookbook internals break.

### Steps to Reproduce:
```
grafana_config_auth 'grafana' do
  proxy_enabled true
end

grafana_datasource 'prometheus' do
  datasource(
    type: 'prometheus',
    url: 'http://localhost:9090',
    access: 'direct'
  )
end

grafana_user 'joe' do
  user(name: 'joe', email: '', password: 'ignored', isAdmin: false)
  action :create
end
```

### Expected Result:
* Datasource 'prometheus' created
* User 'joe' created

### Actual Result:
* grafana_dashboard
  * cookbook: silently passes
   ```
         * grafana_datasource[prometheus] action create
       
           - Creating datasource  prometheus
   ```
  * backend:
  ```
  $ curl -s -H 'X-WEBAUTH-USER: admin' http://localhost:3000/api/datasources | jq .
  []
  ```
* grafana_user
  * cookbook: (get_user_list returns nil)
   ```
           NoMethodError
           -------------
           undefined method `each' for nil:NilClass
           
           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/grafana/resources/user.rb:49:in `block in class_from_file'
  ```
  * User not created
  ```
  $ curl -s -H 'X-WEBAUTH-USER: admin' http://localhost:3000/api/users| jq .
  [
    {
      "id": 1,
      "name": "",
      "login": "admin",
      "email": "admin@localhost",
      "avatarUrl": "/grafana/avatar/46d229b033af06a191ff2267bca9ae56",
      "isAdmin": true,
      "lastSeenAt": "2019-05-14T17:25:37+10:00",
      "lastSeenAtAge": "< 1m"
    }
  ]
  ```

### Discussion
I don't have an elegant solution on how to fix.

Changing the `grafana_endpoint` to the fronting httpd is not suitable as the authentication mechanism may not be grafana POST compatible (eg: saml, oidc).

It would be ideal if do_request would honour `grafana_config_auth` and inject header`"#{proxy_header_name}: #{grafana_options[:user]}"`

At a minimum do_request should probably handle/raise non-2XX responses.

**Comment by atward** [comment]
Some debugging re: how login/session logic changes when injecting header. 

Login + header
```
$ wget -O - --keep-session-cookies --save-cookies cookies.txt --post-data 'user=admin&email=&password=admin' -S --header 'X-WEBAUTH-USER: admin' http://localhost:3000/login
--2019-05-14 18:04:13--  http://localhost:3000/login
Resolving localhost (localhost)... 127.0.0.1, ::1
Connecting to localhost (localhost)|127.0.0.1|:3000... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  Content-Type: application/json
  Set-Cookie: grafana_session=3671a443c52494a5bc5273a8c273e5ea; Path=/; Max-Age=2595600; HttpOnly; SameSite=Lax
  Date: Tue, 14 May 2019 08:04:13 GMT
  Content-Length: 23
Length: 23 [application/json]
Saving to: ‘STDOUT’

 0% [                                                                                                                                                ] 0           --.-K/s              {100%[===============================================================================================================================================>] 23          --.-K/s   in 0s      

2019-05-14 18:04:13 (5.91 MB/s) - written to stdout [23/23]
```
cookie
```
$ cat cookies.txt 
# HTTP cookie file.
# Generated by Wget on 2019-05-14 18:04:13.
# Edit at your own risk.

localhost:3000	FALSE	/	FALSE	1560416653	grafana_session	3671a443c52494a5bc5273a8c273e5ea
```
api using session cookie
```
$ wget -O - --load-cookies cookies.txt -S http://localhost:3000/api/users
--2019-05-14 18:04:49--  http://localhost:3000/api/users
Resolving localhost (localhost)... 127.0.0.1, ::1
Connecting to localhost (localhost)|127.0.0.1|:3000... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  Cache-Control: no-cache
  Content-Type: application/json
  Expires: -1
  Pragma: no-cache
  Date: Tue, 14 May 2019 08:04:49 GMT
  Content-Length: 194
Length: 194 [application/json]
Saving to: ‘STDOUT’

 0% [                                                                                                                                                ] 0           --.-K/s              
[{"id":1,"name":"","login":"admin","email":"admin@localhost","avatarUrl":"/avatar/46d229b033af06a191ff2267bca9ae56","isAdmin":true,"lastSeenAt":"2019-05-14T18:02:33+10:00","lastSeenAtAge100%[===============================================================================================================================================>] 194         --.-K/s   in 0s      

2019-05-14 18:04:49 (53.3 MB/s) - written to stdout [194/194]
```
api without session
```
 wget -S http://localhost:3000/api/users
--2019-05-14 18:08:23--  http://localhost:3000/api/users
Resolving localhost (localhost)... 127.0.0.1, ::1
Connecting to localhost (localhost)|127.0.0.1|:3000... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 401 Unauthorized
  Cache-Control: no-cache
  Content-Type: application/json; charset=UTF-8
  Expires: -1
  Pragma: no-cache
  Date: Tue, 14 May 2019 08:08:23 GMT
  Content-Length: 26
Authorization failed.
```
tl;dr only login needs to inject header. Subsequent request seem to honour session

**Comment by atward** [comment]
Could pass these headers in: https://github.com/sous-chefs/grafana/compare/master...atward:issue-247-poc. 
@Xorima would this be an acceptable approach? If so, I can complete and raise PR

**Comment by xorima** [comment]
If it works and passes the tests please do, I haven't ever looked into the api so will leave this to your judgement 👍
